core_impl/src/main/java/com/google/instrumentation/common/EventQueue.java

```diff
@@ -0,0 +1,165 @@
+ * public class someClass {
+ *   public void doSomething() {
+ *     // Do the work of the method. One result is a measurement of something.
+ *     int measurement = doSomeWork();
+ *     // Make an update to the class state, based on this measurement. This work can take some
+ *     // time, but can be done asynchronously, in the background.
+ *     update(measurement);
+ *   }
+ *
+ *   public void update(int arg) {
+ *     // do something
+ *   }
+ * }
+ * </pre>
+ *
+ * <p>The work of calling {@code someClass.update()} can be executed in the backgound as follows:
+ * <pre>
+ * public class someClass {
+ *   // Add a EventQueueEntry class that will process the update call.
+ *   private static final class SomeClassUpdateEvent implements EventQueueEntry {
+ *     private final SomeClass someClassInstance;
+ *     private final int arg;
+ *
+ *     SomeObjectUpdateEvent(SomeObject someClassInstance, int arg) {
+ *       this.someClassInstance = someClassInstance;
+ *       this.arg = arg;
+ *     }
+ *
+ *     &#064;Override
+ *     public void process() {
+ *       someClassInstance.update(arg);
+ *     }
+ *   }
+ *
+ *   public void doSomething() {
+ *     int measurement = doSomeWork();
+ *     // Instead of calling update() directly, create an event to do the processing, and insert
+ *     // it into the EventQueue. It will be processed in a background thread, and doSomething()
+ *     // can return immediately.
+ *     EventQueue.getInstance.enqueue(new SomeClassUpdateEvent(this, measurement));
+ *   }
+ * }
+ *</pre>
+ */
+
+@ThreadSafe
+public final class EventQueue {
+  // An event in the {@link EventQueue}. Just holds a reference to an EventQueueEntry.
+  private static final class InstrumentationEvent {
+    private EventQueueEntry entry = null;
+
+    // Set the EventQueueEntry associated with this InstrumentationEvent.
+    void setEntry(EventQueueEntry entry) {
+      this.entry = entry;
+    }
+
+    // Get the EventQueueEntry associated with this InstrumentationEvent.
+    EventQueueEntry getEntry() {
+      return entry;
+    }
+  }
+
+  // Factory for InstrumentationEvent.
+  private static final class InstrumentationEventFactory
+      implements EventFactory<InstrumentationEvent> {
+    @Override
+    public InstrumentationEvent newInstance() {
+      return new InstrumentationEvent();
+    }
+  }
+
+  /**
+   * Every event that gets added to {@link EventQueue} will get processed here. Just calls the
+   * underlying process() method.
+   */
+  private static final class InstrumentationEventHandler
+      implements EventHandler<InstrumentationEvent> {
+    @Override
+    public void onEvent(InstrumentationEvent event, long sequence, boolean endOfBatch) {
+      event.getEntry().process();
+    }
+  }
+
+  // The single instance of the class.
+  private static final EventQueue eventQueue = new EventQueue();
+  // The event queue is built on this {@link Disruptor}.
+  private final Disruptor<InstrumentationEvent> disruptor;
```

Comment:
More classes that use it coming soon. Re shading, lets wait and see; I suspect we may want to do this, but if so, will do a separate PR.